### ISSUE TITLE
[BUG] User-defined `__setitem__` method is failing with a "must be mutable" error

### ISSUE THREAD

**Issue by leandrolcampos** [modularml/mojo#3142]
### Bug description

I encountered an issue when trying to assign a value to an element of a custom list structure using the `[]` operator. The assignment fails with an error stating that the expression must be mutable, even though the custom list structure implements the `__setitem__` method.

### Steps to reproduce

Copy and paste the code below into a `main.mojo` file:

```mojo
struct IntList:
    """Just a toy example."""

    var data: List[Int]

    fn __init__(inout self, *values: Int):
        self.data = List[Int](capacity=len(values))
        for value in values:
            self.data.append(value)

    fn __getitem__[idx: Int](self) -> Int:
        return self.data[idx]

    fn __setitem__[idx: Int](inout self, value: Int):
        self.data[idx] = value


fn main():
    var lst = IntList(1, 2, 3, 4)
    lst[0] = 0
    print(lst[0])

```

Then try to execute it:

```
> mojo run ./main.mojo
./main.mojo:20:8: error: expression must be mutable in assignment
    lst[0] = 0
    ~~~^~~
mojo: error: failed to parse the provided Mojo source module
```

The code compile and run successfully if I replace `lst[0] = 0` with `lst.__setitem__[0](0)` or `lst.data[0] = 0`.

### System information

```shell
Ubuntu 22.04.3 LTS
modular 0.8.0 (39a426b5)
mojo 2024.6.2805 (512c667b)
```

**Comment by lattner** [comment]
The error message is not great at all, but the problem is that the compiler cannot resolve the setitem parameter "idx" so it doesn't see the setitem overload.  That's why it is complaining about not being mutable.  I'll take a look at the error message

**Comment by lattner** [comment]
Thank you for reporting this, we now produce:

```
x.mojo:7:8: error: invalid call to '__setitem__': could not deduce parameter 'idx' of callee '__setitem__'
    lst[0] = 0
       ^
x.mojo:4:8: note: function declared here
    fn __setitem__[idx: Int](inout self, value: Int): pass
       ^
```

**Comment by leandrolcampos** [comment]
Hello, @lattner. Thank you so much for taking a look at this.

Considering that you closed the issue, I got the impression that the problem I reported was only with the error message. I couldn't understand why, in this case, it is expected that the compiler cannot resolve the setitem parameter "idx" so it doesn't see the setitem overload. What is the difference between this setitem and, for example, the setitem of the StaticTuple struct?

```mojo
@always_inline("nodebug")
fn __setitem__[index: Int](inout self, val: Self.element_type): ...
```

All the best,
